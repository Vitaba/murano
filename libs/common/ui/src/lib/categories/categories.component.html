<ng-container *ngIf="!errors; else showErrors">
  <div [class]="styles.container">
    <ng-template #imageDefaultTemplate>
      <img [class]="styles.image" [src]="data.image.value"[alt]="data.image.alt"/>
    </ng-template>
    <ng-container [ngTemplateOutlet]="imageExtraTemplate || imageDefaultTemplate">
    </ng-container>


    <ng-template #nameDefaultTemplate>
      <p [class]="styles.name">{{data.name}}</p>
    </ng-template>
    <ng-container [ngTemplateOutlet]="nameExtraTemplate || nameDefaultTemplate">
      </ng-container>
    
  </div>
</ng-container>

<ng-template #showErrors>
    {{ errors | json}}
</ng-template>
  