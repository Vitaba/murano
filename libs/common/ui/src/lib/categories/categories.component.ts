// tslint:disable: no-any no-unsafe-any
import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChild, Input, OnChanges, SimpleChanges, TemplateRef } from '@angular/core';
import { CategoriesData, CategoriesStyles, GuardError, isCategoriesData, isCategoriesStyles } from '@vitaba/common-utils';
import { BehaviorSubject } from 'rxjs';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-categories',
  styleUrls: ['./categories.component.scss'],
  templateUrl: './categories.component.html',
})
export class CategoriesComponent implements AfterViewInit, OnChanges {
  public errors!: GuardError;
  public changes!: BehaviorSubject<SimpleChanges>;
  @Input() public data: CategoriesData = {
    image: {
      alt: 'News Feed Icon',
      value: 'https://www.materialui.co/materialIcons/action/lock_black_144x144.png',
    },
    name: 'News Feed',
  };

  @Input() public styles: CategoriesStyles = {
    container: 'flex',
    image: 'w-6 h-6',
    name: 'text-sm text-gray-600 flex items-center',
  };

  @ContentChild('imageExtraTemplate')
  public imageExtraTemplate!: TemplateRef<any>;

  @ContentChild('nameExtraTemplate')
  public nameExtraTemplate!: TemplateRef<any>;

  public constructor(private readonly _changeDetectorRef: ChangeDetectorRef) {}

  public ngOnChanges(changes: SimpleChanges): void {
    if (!this.changes) {
      this.changes = new BehaviorSubject(changes);
    }
    this.changes.next(changes);
  }

  public ngAfterViewInit(): void {
    this.changes.subscribe((changes: SimpleChanges) => {
      if (changes.data) {
        const validation =
        this.validateCategoriesData(
          changes.data.currentValue);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }
      if (changes.styles) {
        const validation =
        this.validateCategoriesStyles(
          changes.styles.currentValue);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }
      this._changeDetectorRef.detectChanges();
    });
  }

  public validateCategoriesData(value: CategoriesData) {
    const validations = [];

    if (!this.nameExtraTemplate) {
      validations.push({
        name: 'name',
        type: 'string',
      });
    }

    if (!this.imageExtraTemplate) {
      validations.push({
        name: 'image',
        type: 'object',
      });
    }

    const validation = isCategoriesData(value, validations);

    return !validation.valid ? validation : undefined;
  }

  public validateCategoriesStyles(value: CategoriesData) {
    const validations = [{
      name: 'container',
      type: 'string',
    }];

    if (!this.nameExtraTemplate) {
      validations.push({
        name: 'name',
        type: 'string',
      });
    }

    if (!this.imageExtraTemplate) {
      validations.push({
        name: 'image',
        type: 'string',
      });
    }

    const validation = isCategoriesStyles(value, validations);

    return !validation.valid ? validation : undefined;
  }
}
