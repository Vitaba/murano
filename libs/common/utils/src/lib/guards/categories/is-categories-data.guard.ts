import { GuardError } from '../../interfaces/guard/guard.interface';
import { hasProperties } from '../common/has-properties.guard';
import { isObject } from '../common/is-object.guard';

// tslint:disable: no-any no-unsafe-any

export function isCategoriesData(
  arg: any,
  properties: Array<{ name: string; type: string }>,
) {
  const isObjectValidation = isObject(arg);
  const hasPropertiesValidation = hasProperties(arg, properties);
  const valid = isObjectValidation && hasPropertiesValidation.valid;

  const errors = [];

  if (!isObjectValidation.valid) {
    errors.push({ ...isObjectValidation });
  }
  if (!hasPropertiesValidation.valid) {
    errors.push({ ...hasPropertiesValidation });
  }

  const compose: GuardError = {
    valid,
    guard: 'isCategoriesData',
    value: arg,
  };

  if (!valid) {
    compose.message = 'Is not a categories data';
    compose.errors = errors;
  }

  return compose;
}
