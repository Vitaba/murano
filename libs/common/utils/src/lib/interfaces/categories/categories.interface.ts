export interface CategoriesData {
  name: string;
  image: {
    alt: string;
    value: string;
  };
}

export interface CategoriesStyles {
  container: string;
  name: string;
  image: string;
}
