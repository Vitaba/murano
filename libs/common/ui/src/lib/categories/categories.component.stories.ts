// tslint:disable:no-default-export object-literal-sort-keys
import { text } from '@storybook/addon-knobs';
import { CategoriesComponent } from './categories.component';
export default {
  title: 'CategoriesComponent',
};

const data = {
  image: {
    alt: 'Category Icon',
    value: 'https://www.materialui.co/materialIcons/action/lock_black_144x144.png',
  },
  name: 'News Feed',
};

const styles = {
  container: 'flex',
  image: 'w-6 h-6',
  name: 'text-sm text-gray-600 flex items-center',
};

export const basic = () => ({
  moduleMetadata: {
    declarations: [CategoriesComponent],
    imports: [],
  },
  props: {
    data_image_alt: text('image alt', data.image.alt, 'data'),
    data_image_value: text('image value', data.image.value, 'data'),
    data_name: text('name', data.name, 'data'),
    styles_container: text('container', styles.container, 'styles'),
    styles_image: text('image', styles.image, 'styles'),
    styles_name: text('name', styles.name, 'styles'),
  },
  template: `
  <vitaba-categories
  [data]="
  {
    image: {
      alt: data_image_alt,
      value: data_image_value
    },
    name: data_name
  }"
  [styles]="
  {
    container: styles_container,
    image: styles_image,
    name: styles_name
  }
  ">
  </vitaba-categories>`,
  stylesUrl: [
    'node_modules/tailwindcss/dist/base',
    'node_modules/tailwindcss/dist/components',
    'node_modules/tailwindcss/dist/utilities',
  ]
});

export const nameProjection = () => ({
  moduleMetadata: {
    declarations: [CategoriesComponent],
    imports: [],
  },
  props: {
    data_image_alt: text('image alt', data.image.alt, 'data'),
    data_image_value: text('image value', data.image.value, 'data'),
    data_name: text('name', data.name, 'data'),
    styles_container: text('container', styles.container, 'styles'),
    styles_image: text('image', styles.image, 'styles'),
    styles_name: text('name', styles.name, 'styles'),
  },
  template: `
  <vitaba-categories
  [data]="
  {
    image: {
      alt: data_image_alt,
      value: data_image_value
    },
    name: data_name
  }"
  [styles]="
  {
    container: styles_container,
    image: styles_image,
    name: styles_name
  }
  ">
  <ng-template #nameExtraTemplate> <b>Name Extra Template</b> </ng-template>
  </vitaba-categories>`,
  stylesUrl: [
    'node_modules/tailwindcss/dist/base',
    'node_modules/tailwindcss/dist/components',
    'node_modules/tailwindcss/dist/utilities',
  ]
});

export const imageProjection = () => ({
  moduleMetadata: {
    declarations: [CategoriesComponent],
    imports: [],
  },
  props: {
    data_image_alt: text('image alt', data.image.alt, 'data'),
    data_image_value: text('image value', data.image.value, 'data'),
    data_name: text('name', data.name, 'data'),
    styles_container: text('container', styles.container, 'styles'),
    styles_image: text('image', styles.image, 'styles'),
    styles_name: text('name', styles.name, 'styles'),
  },
  template: `
  <vitaba-categories
  [data]="
  {
    image: {
      alt: data_image_alt,
      value: data_image_value
    },
    name: data_name
  }"
  [styles]="
  {
    container: styles_container,
    image: styles_image,
    name: styles_name
  }
  ">
  <ng-template #imageExtraTemplate> <b>Image Extra Template</b> </ng-template>
  </vitaba-categories>`,
  stylesUrl: [
    'node_modules/tailwindcss/dist/base',
    'node_modules/tailwindcss/dist/components',
    'node_modules/tailwindcss/dist/utilities',
  ]
});
