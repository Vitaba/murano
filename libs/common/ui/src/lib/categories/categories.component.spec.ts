
// tslint:disable: no-any component-max-inline-declarations ter-max-len
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { CategoriesComponent } from './categories.component';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-categories-sample-implementation',
  template: `
  <ng-container [ngSwitch]="mode">
    <vitaba-categories *ngSwitchDefault [data]="data" [styles]="styles">
    </vitaba-categories>

    <vitaba-categories *ngSwitchCase="'name'" [data]="data" [styles]="styles">
      <ng-template #nameExtraTemplate>
        <p>Name content-projection</p>
      </ng-template>
    </vitaba-categories>

    <vitaba-categories *ngSwitchCase="'image'" [data]="data" [styles]="styles">
      <ng-template #imageExtraTemplate>
        <p>Image content-projection</p>
      </ng-template>
    </vitaba-categories>
  </ng-container>
  `,
})
class TestHostComponent {
  public data = {
    image: {
      alt: 'News Feed Icon',
      value: 'https://www.materialui.co/materialIcons/action/lock_black_144x144.png',
    },
    name: 'News Feed',
  };
  public styles = {
    container: 'flex',
    image: 'w-6 h-6',
    name: 'text-sm text-gray-600 flex items-center',
  };
  public mode = 'default';
}

describe('CategoriesComponent', () => {
  let component: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestHostComponent, CategoriesComponent],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should be default', () => {
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be imageExtraTemplate', () => {
    component.mode = 'image';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be nameExtraTemplate', () => {
    component.mode = 'name';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should throw errors when data is not in right format', () => {
    component.mode = 'default';
    component.data = {
      image: 1,
      name: 1,
    } as any;
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
