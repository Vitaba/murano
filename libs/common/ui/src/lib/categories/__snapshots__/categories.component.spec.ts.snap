// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesComponent should be default 1`] = `
<vitaba-categories-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-categories>
    
    
    <div
      class="flex"
    >
      
      
      <img
        alt="News Feed Icon"
        class="w-6 h-6"
        src="https://www.materialui.co/materialIcons/action/lock_black_144x144.png"
      />
      
      
      <p
        class="text-sm text-gray-600 flex items-center"
      >
        News Feed
      </p>
    </div>
    
  </vitaba-categories>
</vitaba-categories-sample-implementation>
`;

exports[`CategoriesComponent should be imageExtraTemplate 1`] = `
<vitaba-categories-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-categories>
    
    
    <div
      class="flex"
    >
      
      
      <p>
        Image content-projection
      </p>
      
      
      <p
        class="text-sm text-gray-600 flex items-center"
      >
        News Feed
      </p>
    </div>
    
  </vitaba-categories>
</vitaba-categories-sample-implementation>
`;

exports[`CategoriesComponent should be nameExtraTemplate 1`] = `
<vitaba-categories-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-categories>
    
    
    <div
      class="flex"
    >
      
      
      <img
        alt="News Feed Icon"
        class="w-6 h-6"
        src="https://www.materialui.co/materialIcons/action/lock_black_144x144.png"
      />
      
      
      <p>
        Name content-projection
      </p>
    </div>
    
  </vitaba-categories>
</vitaba-categories-sample-implementation>
`;

exports[`CategoriesComponent should throw errors when data is not in right format 1`] = `
<vitaba-categories-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-categories>
    
     {
    "valid": false,
    "guard": "isCategoriesData",
    "value": {
      "image": 1,
      "name": 1
    },
    "message": "Is not a categories data",
    "errors": [
      {
        "valid": false,
        "guard": "hasProperties",
        "value": {
          "image": 1,
          "name": 1
        },
        "errors": [
          {
            "value": {
              "name": "name",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          },
          {
            "value": {
              "name": "image",
              "type": "object"
            },
            "guard": "Object Validation",
            "message": "Is not an object",
            "valid": false
          }
        ]
      }
    ]
  }
  
    
  </vitaba-categories>
</vitaba-categories-sample-implementation>
`;
